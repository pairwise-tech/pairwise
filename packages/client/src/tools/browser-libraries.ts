// NOTE: raw-loader! (as opposed to !raw-loader!) does not skip other loaders,
// so the TS file _will be_ compiled for us.

// @ts-ignore
import EXPECTATION_LIB from "!raw-loader!../js/browser-test-lib.js";

// @ts-ignore
import EXPECTATION_LIB_TYPES from "!raw-loader!../js/browser-test-lib.d.ts";

// @ts-ignore
import TEST_EXPECTATION_LIB from "!raw-loader!../js/test-expectation-lib.js";

// @ts-ignore
import TEST_EXPECTATION_LIB_TYPES from "!raw-loader!../js/test-expectation-lib.d.ts";

// @ts-ignore
import EXPRESS_JS_LIB from "!raw-loader!../js/express-js-lib.js";

// @ts-ignore
import EXPRESS_JS_LIB_TYPES from "!raw-loader!../js/express-js-lib.d.ts";

// @ts-ignore
import MONACO_TYPE_PATCHES from "!raw-loader!../monaco-types/monaco-type-patches.d.ts";

// @ts-ignore
import REACT_D_TS from "!raw-loader!../monaco-types/react.d.ts";

// @ts-ignore
import REACT_DOM_D_TS from "!raw-loader!../monaco-types/react-dom.d.ts";

const WORKSPACE_LIB = `
  ${EXPECTATION_LIB}
  ${TEST_EXPECTATION_LIB}
`;

const WORKSPACE_LIB_TYPES = `
  ${EXPECTATION_LIB_TYPES}
  ${TEST_EXPECTATION_LIB_TYPES}
`;

/**
 * Isolate this module to isolate the raw-loader from being pulled into
 * the Jest test environment.
 */
export {
  MONACO_TYPE_PATCHES,
  WORKSPACE_LIB,
  WORKSPACE_LIB_TYPES,
  REACT_D_TS,
  REACT_DOM_D_TS,
  TEST_EXPECTATION_LIB_TYPES,
  EXPRESS_JS_LIB,
  EXPRESS_JS_LIB_TYPES,
};
