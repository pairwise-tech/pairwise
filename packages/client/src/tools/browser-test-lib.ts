// NOTE: raw-loader! (as opposed to !raw-loader!) does not skip other loaders,
// so the TS file _will be_ compiled for us.

// @ts-ignore
// eslint-disable-next-line import/no-webpack-loader-syntax
import EXPECTATION_LIB from "raw-loader!../js/browser-test-lib.ts";

// @ts-ignore
// eslint-disable-next-line import/no-webpack-loader-syntax
import EXPECTATION_LIB_TYPES from "!raw-loader!../js/browser-test-lib.d.ts";

/**
 * Isolate this module to isolate the raw-loader from being pulled into
 * the Jest test environment.
 */
export { EXPECTATION_LIB, EXPECTATION_LIB_TYPES };
