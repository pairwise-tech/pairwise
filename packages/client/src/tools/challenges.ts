// Import Workers:
// @ts-ignore
// eslint-disable-next-line import/no-webpack-loader-syntax
import CodeFormatWorker from "workerize-loader!./prettier-code-formatter";

import { CHALLENGE_TYPE } from "@prototype/common";

/** ===========================================================================
 * Types
 * ============================================================================
 */

export interface TestCaseReact {
  message: string;
  test: string;
  testResult: boolean;
}

export interface TestCaseMarkupTypescript {
  test: string;
  message: string;
  testResult: boolean;
}

export type TestCase = TestCaseReact | TestCaseMarkupTypescript;

export enum IFRAME_MESSAGE_TYPES {
  LOG = "LOG",
  INFO = "INFO",
  WARN = "WARN",
  ERROR = "ERROR",
  TEST_RESULTS = "TEST_RESULTS",
  TEST_ERROR = "TEST_ERROR",
}

export interface IframeMessageEvent extends MessageEvent {
  data: {
    message: string;
    source: IFRAME_MESSAGE_TYPES;
  };
}

export interface CodeFormatMessage {
  code: string;
  type: CHALLENGE_TYPE;
  channel: string;
}

export interface CodeFormatMessageEvent extends MessageEvent {
  data: CodeFormatMessage;
}

// NOTE: Instatiating the worker right here at the top level feels off. Could
// potentially cause build issues if we get too tricky with our build, but for
// now this should be fine
const codeWorker = new CodeFormatWorker();

export const requestCodeFormatting = (message: CodeFormatMessage) => {
  codeWorker.postMessage(message);
};

export const subscribeCodeWorker = (fn: (e: CodeFormatMessageEvent) => any) => {
  codeWorker.addEventListener("message", fn);
};

export const unsubscribeCodeWorker = (
  fn: (e: CodeFormatMessageEvent) => any,
) => {
  codeWorker.removeEventListener("message", fn);
};

/** ===========================================================================
 * Challenge Utils
 * ============================================================================
 */

/**
 * Just a simpler (and more declarative) way to create dom elements. We need
 * this for modifying the iframe
 */
export const makeElementFactory = (
  createElement: typeof document.createElement,
) => {
  return (tag: string, props: { [k: string]: string }) => {
    const el = createElement(tag);
    Object.keys(props).forEach(k => {
      const v = props[k];
      // @ts-ignore
      el[k] = v;
    });
    return el;
  };
};

export const waitForDom = (
  doc: typeof window.document,
  timeout: number = 10000,
): Promise<void> => {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(
        new Error(`[DOM Timeout] Waited ${timeout}ms without DOM loaded event`),
      );
    }, timeout);

    const listener = () => {
      // Remove self before resolving
      clearTimeout(timer);
      doc.removeEventListener("DOMContentLoaded", listener);
      resolve();
    };

    doc.addEventListener("DOMContentLoaded", listener);
  });
};

/**
 * Some sample code to run provided tests against a challenge and post
 * the messages back to the app to render.
 *
 * Regular TypeScript challenges export simple functions which can just be
 * tested with expected input/output values.
 */
export const getTestCodeTypeScript = (
  testCases: ReadonlyArray<TestCaseMarkupTypescript>,
) => `
let results = [];

for (const x of ${JSON.stringify(testCases)}) {
  const { input, expected } = x;
  results.push(main(...input) === expected);
}

window.parent.postMessage({
  message: JSON.stringify(results),
  source: "TEST_RESULTS",
});
`;

/**
 * Get the test code string for a markup challenge.
 */
export const getTestHarness = (testCode: string): string => `
window.$ = (...args) => document.querySelector(...args);
window.$$ = (...args) => Array.prototype.slice.call(document.querySelectorAll(...args));
window.getStyle = (el, cssProp) => {
  const view = (el.ownerDocument && el.ownerDocument.defaultView) ? el.ownerDocument.defaultView : window;
  const style = view.getComputedStyle(el);
  return style.getPropertyValue(cssProp) || style[cssProp];
}
const assertEqual = (a, b) => {
  if (a !== b) {
    const typeA = typeof a;
    const typeB = typeof b;
    throw new Error(\`[Assertion Error] Expected \${typeA} argument \${a} to equal \${typeB} argument \${b}\`);
  }
  return true;
}
window.expect = (actual) => ({
  toBe: (expected) => assertEqual(actual, expected),
  toBeTruthy: () => assertEqual(Boolean(actual), true),
  toBeFalsy: () => assertEqual(!Boolean(actual), false),
})

function buildTestsFromCode() {
    const arr = [];
    const test = (message, fn) => {
        arr.push({
            message,
            test: fn,
        })
    }

    ${testCode}

    return arr;
}

function runTests() {
  const tests = buildTestsFromCode()

  const results = tests.reduce((agg, { message, test }) => {
    try {
      const _result = test();
      // Tests that pass using expect will return undefined, since they don't return anything.
      // TODO: At some point we will want to account for async tests, which will require
      // changes here
      const testResult = _result === undefined ? true : _result;
      return agg.concat([{
        message,
        testResult: testResult,
        error: null,
      }])
    } catch (err) {
      return agg.concat([{
        message,
        testResult: false,
        error: err.message,
      }])
    }
  }, []);

  return results;
}

try {
  const results = runTests();
  window.parent.postMessage({
    message: JSON.stringify(results),
    source: "${IFRAME_MESSAGE_TYPES.TEST_RESULTS}",
  });
} catch (err) {
  window.parent.postMessage({
    message: JSON.stringify({
      error: err.message,
    }),
    source: "${IFRAME_MESSAGE_TYPES.TEST_ERROR}",
  });
}
`;

/**
 * Some sample code to run provided tests against a challenge and post
 * the messages back to the app to render.
 *
 * This is currently using the React test-utils package for testing,
 * ref: https://reactjs.org/docs/test-utils.html.
 */
export const getTestCodeReact = (testCases: ReadonlyArray<TestCaseReact>) => {
  const codeString = testCases.map(t => t.test).join(", ");

  return `
  {
    const results = [${codeString}];

    window.parent.postMessage({
      message: JSON.stringify(results),
      source: "TEST_RESULTS",
    });
  }
  `;
};
