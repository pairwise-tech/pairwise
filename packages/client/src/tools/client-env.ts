import getenv from "getenv";

/** ===========================================================================
 * Environment Variables Configuration
 * ============================================================================
 */

export const UNSPECIFIED_ENV_GUARD = "@@INVALID_DEV_ENV_VALUE";

export const NODE_ENV = getenv.string("NODE_ENV", UNSPECIFIED_ENV_GUARD);
export const DEV_MODE = getenv.bool("REACT_APP_DEV", false);

// NOTE: Use https for testing authentication locally:
// export const HOST = getenv.string("REACT_APP_HOST", "https://localhost:9000");
export const HOST = getenv.string("REACT_APP_HOST", "http://localhost:9000");

export const CODEPRESS_HOST = getenv.string(
  "REACT_APP_CODEPRESS_HOST",
  "http://localhost:3001",
);

export const TEST = NODE_ENV === "test";
export const DEVELOPMENT = NODE_ENV === "development";
export const PRODUCTION = NODE_ENV === "production";
