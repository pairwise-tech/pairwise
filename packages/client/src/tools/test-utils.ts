import { pipe } from "ramda";
import * as Babel from "@babel/standalone";

import DependencyCacheService from "./dependency-service";
import { Challenge, CHALLENGE_TYPE } from "@pairwise/common";
import protect from "../js/loop-protect-lib.js";
import { TEST } from "./client-env";

/** ===========================================================================
 * Types & Config
 * ============================================================================
 */

export enum IFRAME_MESSAGE_TYPES {
  LOG = "LOG",
  INFO = "INFO",
  WARN = "WARN",
  ERROR = "ERROR",
  TEST_RESULTS = "TEST_RESULTS",
  TEST_ERROR = "TEST_ERROR",
  INFINITE_LOOP = "INFINITE_LOOP",
}

export interface IframeMessageEvent extends MessageEvent {
  data: {
    message: string;
    source: IFRAME_MESSAGE_TYPES;
  };
}

export interface TestCase {
  test: string;
  message: string;
  testResult: boolean;
  error?: string;
}

export interface CodeFormatMessage {
  code: string;
  type: CHALLENGE_TYPE;
  channel: string;
}

export interface CodeFormatMessageEvent extends MessageEvent {
  data: CodeFormatMessage;
}

/**
 * Functions used to intercept console methods and post the messages to
 * the parent window.
 */
const CONSOLE_INTERCEPTOR_FUNCTIONS = `
const __replacer = (key, value) => {
  if (typeof value === "undefined") {
    return "__transform_undefined__";
  }

  return value;
}

const __interceptConsoleLog = (...value) => {
  window.parent.postMessage({
    message: JSON.stringify(value, __replacer),
    source: "LOG",
  });
}

const __interceptConsoleInfo = (...value) => {
  window.parent.postMessage({
    message: JSON.stringify(value, __replacer),
    source: "INFO",
  });
}

const __interceptConsoleWarn = (...value) => {
  window.parent.postMessage({
    message: JSON.stringify(value, __replacer),
    source: "WARN",
  });
}

const __interceptConsoleError = (...value) => {
  window.parent.postMessage({
    message: JSON.stringify(value, __replacer),
    source: "ERROR",
  });
}
`;

const INFINITE_LOOP_TIMEOUT = 2000;

/**
 * Register loop-protect Babel plugin.
 */
Babel.registerPlugin(
  "loopProtection",
  protect(INFINITE_LOOP_TIMEOUT, () => {
    throw new Error("INFINITE_LOOP");
  }),
);

/** ===========================================================================
 * Test Utils
 * ============================================================================
 */

/**
 * This function is supposed to match all import statements in the code string
 * and remove them, while also identifying the imported libraries and returning
 * those in an array by name, so they can be fetched from a CDN and injected
 * into the code before it is transpiled and run.
 *
 * The code may not work 100%:
 */
export const stripAndExtractModuleImports = (codeString: string) => {
  // Reference: https://gist.github.com/manekinekko/7e58a17bc62a9be47172
  const regex = new RegExp(
    /import(?:["'\s]*([\w*{}\n\r\t, ]+)from\s*)?["'\s].*([@\w/_-]+)["'\s].*/g,
  );
  const result = codeString.match(regex);

  let strippedImports = codeString;
  let dependencies: ReadonlyArray<string> = [];

  if (result) {
    for (const importStatement of result) {
      const libs = importStatement.match(/"(.*?)"/);
      if (libs) {
        dependencies = dependencies.concat(libs[1]);
      }

      strippedImports = strippedImports.replace(importStatement, "");
    }
  }

  return {
    dependencies,
    code: strippedImports,
  };
};

/**
 * Replace all console.log statements with a call to a custom function which
 * is injected on the top of the code string before it is run. The custom
 * function will post a message outside of the iframe to the parent window
 * object, which is listening to capture these messages and serve them to the
 * workspace console.
 */
export const hijackConsole = (codeString: string) => {
  const replacedConsole = codeString
    .replace(/console.log/g, "__interceptConsoleLog")
    .replace(/console.info/g, "__interceptConsoleInfo")
    .replace(/console.warn/g, "__interceptConsoleWarn")
    .replace(/console.error/g, "__interceptConsoleError");

  return `${CONSOLE_INTERCEPTOR_FUNCTIONS}${replacedConsole}`;
};

/**
 * Remove all calls to console methods in a code string. This just removes
 * them with // to create a comment which will not be evaluated... this could
 * create problems if the console statement was written inline with some
 * real code, which would then no longer execute...
 */
export const stripConsoleCalls = (codeString: string) => {
  return codeString
    .replace(/console.log/g, "// ")
    .replace(/console.info/g, "// ")
    .replace(/console.warn/g, "// ")
    .replace(/console.error/g, "// ");
};

/**
 * Transpile the code use Babel standalone module.
 */
export const transpileCodeWithBabel = (codeString: string) => {
  /**
   * SUFFER!
   *
   * For some reason I couldn't get babel-standalone to transform class
   * properties which would break React code in the test environment. Instead
   * I just to the following hideous thing.
   *
   * TODO: Fix it and find a way to run everything through babel-standalone.
   */
  if (TEST) {
    return require("babel-core").transform(codeString, {
      presets: [
        "@babel/preset-react",
        ["@babel/preset-typescript", { isTSX: true, allExtensions: true }],
      ],
      plugins: ["@babel/plugin-proposal-class-properties"],
    }).code;
  } else {
    return Babel.transform(codeString, {
      presets: [
        "es2017",
        "react",
        ["typescript", { isTSX: true, allExtensions: true }],
      ],
      plugins: ["loopProtection"],
    }).code;
  }
};

/**
 * Get all imported dependencies from the DependencyCacheService.
 */
const fetchRequiredDependencies = async (
  dependencies: ReadonlyArray<string>,
) => {
  return Promise.all(dependencies.map(DependencyCacheService.getDependency));
};

/**
 * Concatenate package source dependencies to code string.
 */
const injectDependencies = (
  codeString: string,
  dependencies: readonly string[],
) => {
  let result = "";

  for (const dependency of dependencies) {
    result += `${dependency}\n`;
  }

  result += codeString;
  return result;
};

/**
 * Fetch the required module dependencies and inject them into the code string.
 */
export const createInjectDependenciesFunction = (
  dependencies: readonly string[],
) => async (codeString: string) => {
  /**
   * TODO: The following method could throw an error if an imported package
   * cannot be found, this should be handled somewhere.
   */
  const dependencySourceList = await fetchRequiredDependencies(dependencies);
  const codeWithDependencies = injectDependencies(
    codeString,
    dependencySourceList,
  );

  return codeWithDependencies;
};

/**
 * Inject test code into a code string.
 *
 * The test string is injected twice. The first injection runs normally
 * and produces the output the user will see. The second injection is
 * enclosed in a new scope and combined with the test code, the console
 * is also modified and removed.
 *
 * Together, this creates one code string which is executed to produce the
 * preview for the user and the test results for the workspace.
 */
export const injectTestCode = (testCode: string) => (codeString: string) => {
  return `
    /* Via injectTestCode */
    {
      try {
        ${codeString}
      } catch (err) {
        if (err.message === "INFINITE_LOOP") {
          console.error("Infinite loop detected");
        }
      }
    }
    {
      ${getTestHarness(stripConsoleCalls(codeString), testCode)}
    }
    `;
};

/**
 * Get the full html content string for the iframe, injected the user code
 * into it. This currently includes script libraries now.
 *
 * NOTE: I'd like to unify the markdown and ts tests so that when building the
 * markup they would both use some common function like this, but we're not
 * there yet.
 */
export const getMarkupForCodeChallenge = (
  scriptString: string,
  deps: string,
) => `
<html>
  <head></head>
  <body>
    <div id="root" />
    <script id="test-dependencies">${getTestDependencies(deps)}</script>
    <script id="test-code">${scriptString}</script>
  </body>
</html>
`;

/**
 * This is just exported as a function for consistency and in case we need to
 * augment it later.
 *
 * NOTE: This function takes the dependency library as an input. This is a
 * workaround to allow us to inject the dependency library at runtime, because
 * this method runs in the Jest environment and the app, but for the app
 * the dependency file is imported with a raw-loader which is incompatible
 * with Jest.
 */
export const getTestDependencies = (testLib: string): string => testLib;

/**
 * Get the test code string for a markup challenge.
 */
export const getTestHarness = (userCode: string, testCode: string): string => `
try {
  ${userCode}

  function buildTestsFromCode() {
    const testArray = [];
    const test = (message, fn) => {
      testArray.push({
            message,
            test: fn,
        })
    }

    ${testCode}

    return testArray;
  }

  function runTests() {
    const tests = buildTestsFromCode()

    const results = tests.reduce((agg, { message, test }) => {
      try {
        const _result = test();
        // Tests that pass using expect will return undefined, since they don't return anything.
        // TODO: At some point we will want to account for async tests, which will require
        // changes here
        const testResult = _result === undefined ? true : _result;
        return agg.concat([{
          message,
          testResult: testResult,
          error: null,
        }])
      } catch (err) {
        return agg.concat([{
          message,
          testResult: false,
          error: err.message + '\\n\\n' + err.stack,
        }])
      }
    }, []);

    return results;
  }

  try {
    const results = runTests();
    window.parent.postMessage({
      message: JSON.stringify(results),
      source: "${IFRAME_MESSAGE_TYPES.TEST_RESULTS}",
    });
  } catch (err) {
    window.parent.postMessage({
      message: JSON.stringify({
        error: err.message,
      }),
      source: "${IFRAME_MESSAGE_TYPES.TEST_ERROR}",
    });
  }
} catch (err) {
  if (err.message === "INFINITE_LOOP") {
    window.parent.postMessage({
      message: JSON.stringify({
        error: err.message,
      }),
      source: "${IFRAME_MESSAGE_TYPES.INFINITE_LOOP}",
    });
  }

  // compilation error, or something else happened
  // TODO: Still propagate this message back to the Workspace?
}
`;

/**
 * Put together the script tags necessary for running the tests in an iframe,
 * including the script that includes the tests themselves.
 */
export const getTestScripts = (testCode: string, testLib: string) => {
  return `
    <script id="test-code">${getTestHarness(
      getTestDependencies(testLib),
      testCode,
    )}</script>
  `;
};

/**
 * Uses the browsers own internal engine to clean up HTML. The reason to do this
 * is to standardize HTML output from the user before work work with it
 * internally
 */
export const tidyHtml = (html: string) => {
  const el = document.createElement("html");
  el.innerHTML = html;
  return el.innerHTML;
};

/**
 * Compile and process the code string for any challenge. Return compiled code
 * which can be run in the tests.
 *
 * NOTE: This function is async because createInjectDependenciesFunction
 * returns an async function.
 *
 * TODO: Refactor and unify the markup challenge with the other regular
 * code challenges.
 */
export const compileCodeString = async (
  sourceCodeString: string,
  challenge: Challenge,
) => {
  if (challenge.type === "markup") {
    const testScript = getTestScripts(challenge.testCode, "");

    // NOTE: Tidy html should ensure there is indeed a closing body tag
    const tidySource = tidyHtml(sourceCodeString);

    // Just to give us some warning if we ever hit this. Should be impossible...
    if (!tidySource.includes("</body>")) {
      console.warn(
        "[Err] Could not append test code to closing body tag in markup challenge",
      );
    }

    /**
     * TODO: There's no reason for us to inject the test script in sandbox
     * mode, but the same applies to all challenge types so ideally we
     * would standardize the testing pipeline to the point where we could
     * include that logic in one place only.
     */
    const sourceDocument = tidySource.replace(
      "</body>",
      `${testScript}</body>`,
    );

    return { code: sourceDocument, dependencies: [""] };
  } else {
    const { code: sourceCode, dependencies } = stripAndExtractModuleImports(
      sourceCodeString,
    );

    const injectModuleDependenciesFn = createInjectDependenciesFunction(
      challenge.type === "react"
        ? [...dependencies, "react-dom-test-utils"]
        : dependencies,
    );

    /**
     * What happens here:
     *
     * - Inject test code in code string, and remove any console methods
     * - Hijack all console usages in user code string
     * - Apply Babel transform steps
     * - Fetch and inject required modules into code string
     */
    const processedCodeString = await pipe(
      injectTestCode(challenge.testCode),
      hijackConsole,
      transpileCodeWithBabel,
      injectModuleDependenciesFn,
    )(sourceCode);

    return {
      dependencies,
      code: processedCodeString,
    };
  }
};
