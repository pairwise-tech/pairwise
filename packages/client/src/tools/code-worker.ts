// Import Workers:
// @ts-ignore
// eslint-disable-next-line import/no-webpack-loader-syntax
import CodeFormatWorker from "workerize-loader!./prettier-code-formatter";

import { CodeFormatMessage, CodeFormatMessageEvent } from "./test-utils";

/**
 * NOTE: Instantiating the worker right here at the top level feels off. Could
 * potentially cause build issues if we get too tricky with our build, but for
 * now this should be fine.
 */
const codeWorker = new CodeFormatWorker();

export const requestCodeFormatting = (message: CodeFormatMessage) => {
  codeWorker.postMessage(message);
};

export const subscribeCodeWorker = (fn: (e: CodeFormatMessageEvent) => any) => {
  codeWorker.addEventListener("message", fn);
};

export const unsubscribeCodeWorker = (
  fn: (e: CodeFormatMessageEvent) => any,
) => {
  codeWorker.removeEventListener("message", fn);
};
