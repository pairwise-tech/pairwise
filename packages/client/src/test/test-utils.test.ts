/* eslint-disable */
import { buildPreviewTestResultsFromCode } from "tools/test-utils";

describe("test test-utils.ts functions", () => {
  test("buildPreviewTestResultsFromCode successfully builds preview test cases for synchronous test code", () => {
    const testCode =
      // tslint:disable-next-line
      'test("A function `everyNth` should be defined.", () => {\n  expect(everyNth).toBeDefined();\n  expect(typeof everyNth === "function").toBe(true);\n});\n\ntest("`everyNth` should return `[]` given the input `[1,2,3,4]` and `7`.", () => {\n  expect(everyNth([1, 2, 3, 4], 7)).toEqual([]);\n});\n\ntest("`everyNth` should return `[]` given the input `[]` and `7`.", () => {\n  expect(everyNth([], 1)).toEqual([]);\n});\n\ntest("`everyNth` should return `[3,6,9]` given the input `[1,2,3,4,5,6,7,8,9,10]` and `3`.", () => {\n  expect(everyNth([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3)).toEqual([3, 6, 9]);\n});\n\ntest("`everyNth` should return `[5]` given the input `[1,2,3,4,5]` and `5`.", () => {\n  expect(everyNth([1, 2, 3, 4, 5], 5)).toEqual([5]);\n});\n\ntest("`everyNth` should return `[1,2,3,4,5,6,7,8]` given the input `[1,2,3,4,5,6,7,8]` and `1`.", () => {\n  expect(everyNth([1, 2, 3, 4, 5, 6, 7, 8], 1)).toEqual([\n    1,\n    2,\n    3,\n    4,\n    5,\n    6,\n    7,\n    8,\n  ]);\n});\n\ntest("`everyNth` should return `[10]` given the input `[1,2,3,4,5,6,7,8,9,10]` and `10`.", () => {\n  expect(everyNth([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 10)).toEqual([10]);\n});\n';
    const result = buildPreviewTestResultsFromCode(testCode);
    expect(result).toMatchSnapshot();
  });

  test("buildPreviewTestResultsFromCode successfully builds preview test cases for asynchronous test code", () => {
    const testCode =
      // tslint:disable-next-line
      'test("The `SpaceX` class should be defined.", () => {\n  const x = new SpaceX();\n  expect(x).toBeDefined();\n});\n\ntest("The `SpaceX` class should implement the `SpaceXClass` interface and contain all of its methods.", () => {\n  const x = new SpaceX();\n  expect(typeof x.getCrew).toBe("function");\n  expect(typeof x.getLaunchpads).toBe("function");\n  expect(typeof x.getLandpads).toBe("function");\n  expect(typeof x.getRockets).toBe("function");\n  expect(typeof x.getShips).toBe("function");\n  expect(typeof x.getStarlink).toBe("function");\n});\n\nconst __fetchHelper = async (path) => {\n  const response = await fetch(`https://api.spacexdata.com/v4/${path}`);\n  const data = await response.json();\n  return data;\n};\n\nconst __compareHelper = (a, b) => {\n  expect(Object.keys(a)).toEqual(Object.keys(b));\n}\n\ntest("The `getCrew` method should return the data from call the `/crew` endpoint.", async () => {\n  const expected = await __fetchHelper("crew");\n  const x = new SpaceX();\n  const result = await x.getCrew();\n  __compareHelper(expected, result);\n});\n\ntest("The `getLaunchpads` method should return the data from call the `/launchpads` endpoint.", async () => {\n  const expected = await __fetchHelper("launchpads");\n  const x = new SpaceX();\n  const result = await x.getLaunchpads();\n  __compareHelper(expected, result);\n});\n\ntest("The `getLandpads` method should return the data from call the `/landpads` endpoint.", async () => {\n  const expected = await __fetchHelper("landpads");\n  const x = new SpaceX();\n  const result = await x.getLandpads();\n  __compareHelper(expected, result);\n});\n\ntest("The `getRockets` method should return the data from call the `/rockets` endpoint.", async () => {\n  const expected = await __fetchHelper("rockets");\n  const x = new SpaceX();\n  const result = await x.getRockets();\n  __compareHelper(expected, result);\n});\n\ntest("The `getShips` method should return the data from call the `/ships` endpoint.", async () => {\n  const expected = await __fetchHelper("ships");\n  const x = new SpaceX();\n  const result = await x.getShips();\n  __compareHelper(expected, result);\n});\n\ntest("The `getStarlink` method should return the data from call the `/starlink` endpoint.", async () => {\n  const expected = await __fetchHelper("starlink");\n  const x = new SpaceX();\n  const result = await x.getStarlink();\n  __compareHelper(expected, result);\n});\n';
    const result = buildPreviewTestResultsFromCode(testCode);
    expect(result).toMatchSnapshot();
  });

  test("buildPreviewTestResultsFromCode successfully builds preview test cases for one-liner tests", () => {
    const testCode =
      // tslint:disable-next-line
      'test("The log message text in the `setTimeout` should change `1` to `3`.", async () => { const result = await checkBoxAsync(__secret_log_box, `["You will see me after a 3 second delay!"]`, 3500); expect(result).toBeTruthy();});\n\ntest("The `setTimeout` delay should be changed from `1000` milliseconds to `3000` milliseconds.", async () => { expect(__user_code_string__.includes(3000)).toBeTruthy(); const result = await checkBoxAsync(__secret_log_box, `["You will see me after a 3 second delay!"]`, 2000); expect(result).toBeFalsy(); });';
    const result = buildPreviewTestResultsFromCode(testCode);
    expect(result).toMatchSnapshot();
  });

  test("buildPreviewTestResultsFromCode returns an error if testCode does not match regex", () => {
    const testCode = "notest";
    const result = buildPreviewTestResultsFromCode(testCode);
    expect(result).toMatchSnapshot();
  });

  test("buildPreviewTestResultsFromCode handles escaped quotes correctly", () => {
    // tslint:disable-next-line
    const testCode = 'test("Hey there \\" you are amazing", () => {...})';
    const result = buildPreviewTestResultsFromCode(testCode);
    expect(result).toMatchSnapshot();
  });

  test("buildPreviewTestResultsFromCode handles single quotes correctly", () => {
    // tslint:disable-next-line
    const testCode =
      "test('This test case uses single quotes...', () => {...})'";
    const result = buildPreviewTestResultsFromCode(testCode);
    expect(result).toMatchSnapshot();
  });
});
