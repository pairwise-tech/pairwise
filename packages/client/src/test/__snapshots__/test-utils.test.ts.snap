// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test test-utils.ts functions buildPreviewTestResultsFromCode returns an error if testCode does not match regex 1`] = `
Object {
  "error": [Error: Unable to extract test messages from test code!],
}
`;

exports[`test test-utils.ts functions buildPreviewTestResultsFromCode successfully builds preview test cases for asynchronous test code 1`] = `
Object {
  "results": Array [
    Object {
      "message": "The \`SpaceX\` class should be defined.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`SpaceX\` class should implement the \`SpaceXClass\` interface and contain all of its methods.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`getCrew\` method should return the data from call the \`/crew\` endpoint.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`getLaunchpads\` method should return the data from call the \`/launchpads\` endpoint.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`getLandpads\` method should return the data from call the \`/landpads\` endpoint.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`getRockets\` method should return the data from call the \`/rockets\` endpoint.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`getShips\` method should return the data from call the \`/ships\` endpoint.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`getStarlink\` method should return the data from call the \`/starlink\` endpoint.",
      "test": "",
      "testResult": false,
    },
  ],
}
`;

exports[`test test-utils.ts functions buildPreviewTestResultsFromCode successfully builds preview test cases for one-liner tests 1`] = `
Object {
  "results": Array [
    Object {
      "message": "The log message text in the \`setTimeout\` should change \`1\` to \`3\`.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "The \`setTimeout\` delay should be changed from \`1000\` milliseconds to \`3000\` milliseconds.",
      "test": "",
      "testResult": false,
    },
  ],
}
`;

exports[`test test-utils.ts functions buildPreviewTestResultsFromCode successfully builds preview test cases for synchronous test code 1`] = `
Object {
  "results": Array [
    Object {
      "message": "A function \`everyNth\` should be defined.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "\`everyNth\` should return \`[]\` given the input \`[1,2,3,4]\` and \`7\`.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "\`everyNth\` should return \`[]\` given the input \`[]\` and \`7\`.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "\`everyNth\` should return \`[3,6,9]\` given the input \`[1,2,3,4,5,6,7,8,9,10]\` and \`3\`.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "\`everyNth\` should return \`[5]\` given the input \`[1,2,3,4,5]\` and \`5\`.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "\`everyNth\` should return \`[1,2,3,4,5,6,7,8]\` given the input \`[1,2,3,4,5,6,7,8]\` and \`1\`.",
      "test": "",
      "testResult": false,
    },
    Object {
      "message": "\`everyNth\` should return \`[10]\` given the input \`[1,2,3,4,5,6,7,8,9,10]\` and \`10\`.",
      "test": "",
      "testResult": false,
    },
  ],
}
`;
